core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -1238,7 +1238,14 @@ class Partition(val topicPartition: TopicPartition,
   /**
    * Try to complete any pending requests. This should be called without holding the leaderIsrUpdateLock.
    */
-  private def tryCompleteDelayedRequests(): Unit = delayedOperations.checkAndCompleteAll()
+  def tryCompleteDelayedRequests(): Unit = {
```

Comment:
One thing I wasn't sure about:

In the ZK path, we call makeLeaders from becomeLeaderOrFollower/LeaderAndIsrRequest handling. I believe if we encounter an error then, we stop handling the request and the error is handled in KafkaApis. https://github.com/apache/kafka/blob/f1ef7a5a9f87cda7ae1209fc2285259eba4204ca/core/src/main/scala/kafka/server/KafkaApis.scala#L274. BrokerMetadataPublisher's handling is https://github.com/apache/kafka/blob/f1ef7a5a9f87cda7ae1209fc2285259eba4204ca/core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala#L136

With this change, we will continue the makeLeader path in both cases. Perhaps this is safe, but I want to make sure there isn't any gotchas there.

Reply:
The issue is that the delayed requests are just semi-related pieces of code that were waiting for a specific partition transition. It doesn't make sense for them to block the partition transition (and indeed, they can't, either before or after this PR -- the transition has already happened on the controller side.) So throwing an exception here just makes a mess, there are no upsides as far as I can see. Either in ZK mode or KRaft mode.

Reply:
Got it -- so we are improving this code path for both. (But I guess ZK is no more soon anyway 😉 )